src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -172,23 +173,34 @@ private int[] getFilterIdsArray(final LeafReaderContext context) throws IOExcept
         if (filterWeight == null) {
             return new int[0];
         }
-        final BitSet filteredDocsBitSet = getFilteredDocsBitSet(context, this.filterWeight);
-        final int[] filteredIds = new int[filteredDocsBitSet.cardinality()];
-        int filteredIdsIndex = 0;
+        return bitSetToIntArray(getFilteredDocsBitSet(context, this.filterWeight));
+    }
+
+    private int[] getParentIdsArray(final LeafReaderContext context, final BitSetProducer parentFilter) throws IOException {
+        if (parentFilter == null) {
+            return null;
+        }
+        return bitSetToIntArray(parentFilter.getBitSet(context));
+    }
+
+    private int[] bitSetToIntArray(final BitSet bitSet) {
+        final int[] intArray = new int[bitSet.cardinality()];
+        int index = 0;
         int docId = 0;
-        while (docId < filteredDocsBitSet.length()) {
-            docId = filteredDocsBitSet.nextSetBit(docId);
+        while (docId < bitSet.length()) {
```

Comment:
Should we also assert somewhere that index < bitSet.cardinality? Or can we just use a `BitSetIterator` for this loop so that we dont have to deal directly with incrementing docId?

Reply:
no need. bitSet.cardinality is total number of set bit(number of 1 in bits). check of index < bitSet.cardinality doesn't make sense.

Logic is simple but I can utilize BitSetIterator. Let me update the code.